### PR TITLE
Minor: Refactor group by state machine

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Follow on to https://github.com/apache/arrow-datafusion/pull/6932

# Rationale for this change
I was a little unhappy about the `extract_ok!` macro and @tustvold 
offered a suggestion on how to remove it:  https://github.com/apache/arrow-datafusion/pull/6932/files#r1267309181

This PR is what I came up with . I am not really sure it is better but wanted to put it up in case others had an opinion

# What changes are included in this PR?
Pull out the state machine transition logic in group by into some new functions and remove the `extract_ok!` macro

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 1]
```diff
@@ -322,16 +322,6 @@ fn create_group_accumulator(
     }
 }
 
-/// Extracts a successful Ok(_) or returns Poll::Ready(Some(Err(e))) with errors
-macro_rules! extract_ok {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:326]: This macro was removed, but I think breaking the main state loop change into two different functions makes is slightly harder to follow the logic.

Thus I am inclined not to pursue this PR anymore

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 2]
```diff
@@ -342,69 +332,92 @@ impl Stream for GroupedHashAggregateStream {
...
+                let timer = elapsed_compute.timer();
+                let batch = self.create_batch_from_map(EmitTo::All)?;
+                self.exec_state = ExecutionState::ProducingOutput(batch);
+                timer.done();
             }
         }
+        Ok(())
+    }
+
+    /// Produces the next batch of output from ExecutionState::ProductingOuptut, updating self.execution_state
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:398]:
```suggestion
    /// Produces the next batch of output from ExecutionState::ProducingOuptut, updating self.execution_state
```

**Review by tustvold — APPROVED** [review]
I happen to like that this splits out the transition functions from the state machine proper, but I don't feel strongly

**Comment by alamb** [comment]
I think I am inclined to leave this code alone for now as it has already undergone a bunch of churn and this is just stylistic and I want to move on to other things. If anyone else feels differently, please feel free to open a new PR

**Pull request closed without merging**